sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java

```diff
@@ -216,28 +215,54 @@ private HttpResponse attemptSync(final HttpPipelineCallContext context, final Ht
             return httpResponse;
         }
     }
-    private boolean shouldRetry(HttpResponse response, int tryCount) {
+    private static boolean shouldRetry(RetryStrategy retryStrategy, HttpResponse response, int tryCount) {
         return tryCount < retryStrategy.getMaxRetries() && retryStrategy.shouldRetry(response);
     }
 
-    private boolean shouldRetryException(Throwable throwable, int tryCount) {
-        return tryCount < retryStrategy.getMaxRetries() && retryStrategy.shouldRetryException(throwable);
+    private static boolean shouldRetryException(RetryStrategy retryStrategy, Throwable throwable, int tryCount) {
+        // Check if there are any retry attempts still available.
+        if (tryCount >= retryStrategy.getMaxRetries()) {
+            return false;
+        }
+
+        // Unwrap the throwable.
+        Throwable unwrappedThrowable = Exceptions.unwrap(throwable);
+
+        // Check if the unwrapped throwable should be retried.
+        if (retryStrategy.shouldRetryException(unwrappedThrowable)) {
+            return true;
+        }
+
+        // Check if the cause for the throwable should be retried.
+        Throwable causalThrowable = unwrappedThrowable.getCause();
+
+        // Check all causal exceptions in the exception chain.
+        while (causalThrowable != null) {
+            if (retryStrategy.shouldRetryException(causalThrowable)) {
+                return true;
+            }
+
+            causalThrowable = causalThrowable.getCause();
+        }
```

Comment:
Can we do this instead?

```suggestion
        Throwable causalThrowable = Exceptions.unwrap(throwable);

        // Check all causal exceptions in the exception chain.
        while (causalThrowable != null) {
            if (retryStrategy.shouldRetryException(causalThrowable)) {
                return true;
            }

            causalThrowable = causalThrowable.getCause();
        }
```